rfcs/text/0000-metadata-integration-test.md

```diff
@@ -0,0 +1,81 @@
+1. Make failures due to metadata schema changes more readily observable
+1. Increase availability of DCP systems downstream of DSS
+
+## Detailed Design
+
+In this design we will use the following terms:
+ 
+ * "schema integration test system" - an automated CI/CD pipeline used to check that downstream systems can handle bundles with a new schema
+ * "new schema label" - anything that indicates that the schema has changes that have not yet been tested against downstream systems: potentially a JSON field, git tag, git branch, or comparison to tested version list
+
+
+Design
+
+1. Data files ingested with new metadata schema changes are uploaded to the production upload service with a new schema label
+1. Observing the new schema label, the DSS stores the new bundles and makes them available with the `GET /v1/bundles` endpoint with the bundle `uuid` or `(uuid, version)`, but does not index nor does it release subscription notifications for them by default.<sup>1</sup>
+1. Via a schema integration test system, a sample of the new bundles in the production DSS are copied to the integration DSS testing environment. Bundles in the sample are selected for unique experimental graph shapes and fields that may have changed.
+1. The integration DSS deployment stores the bundles, indexes their metadata, and releases bundle events to downstream systems in the integration environment with subscriptions.
+1. The schema integration test checks the results and passes if systems downstream of the integration DSS have correctly processed the new bundles; it fails otherwise.<sup>2</sup>
+1. If the schema integration test passes, the new bundles in the production DSS can be released downstream by issuing a new bundle version of each without the new schema label. This process could be automated.
+1. If the schema test does not pass, the data is not released and development teams are notified to resolve the issue and rerun the test.
+
+![Integration test design](../images/0000-integration-test.png)
+
+*Integration test data flow: test path is shown in red, production release in blue, integration tooling in dashed lines; edges are labeled with corresponding step numbers in the design above*
+
+1. The DSS subscription API should enable clients to opt into receiving events with unsupported metadata schema versions. Notably, this would enable the data browser to display data with unsupported metadata schema versions even if analysis cannot yet process it.
+1. To estimate the impact of schema changes on third party applications, the schema integration test system could run critical [data consumer vignettes](https://github.com/HumanCellAtlas/data-consumer-vignettes).
+
+### Pros:
+
+* Data files only have to be uploaded once
+* Data files can be uploaded directly to the production environment
+* Tests exactly what data downstream systems will see; does not require complicated fake metadata generation code in tests
```

Comment:
I'm concerned that using real data in the integration test may not fully test new schema changes. I know programmatically generated metadata isn't ideal but it allows us to have data for all metadata fields whereas the majority of real datasets only supply metadata for a subset of the available metadata fields. It might therefore be possible that a metadata field change isn't actively used for weeks or even months if the change is made following early discussions between wranglers and submitters, but the actual data submission is held up for any of a number of good reasons.

Reply:
Good point. I think it would be totally reasonable for the implementer of the test to add generation or templating code to more fully test the schema changes. The point here was that we could get value/coverage from the test even without such a system.

Let's leave the final decision on the sampling strategy to the implementer.

Importantly, no integration test will be perfect here. If we can get the 80 from the 80/20 split here, we'd be ahead.